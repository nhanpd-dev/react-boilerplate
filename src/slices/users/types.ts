import { UserModel } from 'models';

/* --- STATE --- */
export interface UsersState {
  loading: boolean;
  error?: RepoErrorType | null;
  users: UserModel[];
}

export enum RepoErrorType {
  RESPONSE_ERROR = 1,
  USER_NOT_FOUND = 2,
  USERNAME_EMPTY = 3,
  USER_HAS_NO_REPO = 4,
  GITHUB_RATE_LIMIT = 5,
}

/* 
  If you want to use 'ContainerState' keyword everywhere in your feature folder, 
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = UsersState;
