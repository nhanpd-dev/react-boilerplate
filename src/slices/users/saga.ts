import { call, takeLatest, put } from 'redux-saga/effects';
import { request } from 'utils/request';
import { actions } from '.';
import { UserModel } from 'models';

export function* getUsers() {
  const requestURL = `http://localhost:3001/users`;

  try {
    const users: UserModel[] = yield call(request, requestURL);
    yield put(actions.usersLoaded(users));
  } catch (err: any) {
    console.log('=====> errors: ', err);
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* usersSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadUsers.type, getUsers);
}
