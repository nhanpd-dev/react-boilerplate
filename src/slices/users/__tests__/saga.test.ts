import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '..';

import { usersSaga, getUsers } from '../saga';
import { RepoErrorType } from '../types';

describe('getUsers Saga', () => {
  let username: any;
  let users: any;
  let getUsersIterator: ReturnType<typeof getUsers>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getUsersIterator = getUsers();
    const delayDescriptor = getUsersIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getUsersIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if username is empty', () => {
    username = '';
    const putDescriptor = getUsersIterator.next(username).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.repoError(RepoErrorType.USERNAME_EMPTY)),
    );

    const iteration = getUsersIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    username = 'test';
    users = [
      {
        id: 1,
        email: 'dainhanphan.dev@yopmail.com',
        password: '12345678q',
        name: 'Dai Nhan Phan',
        birthday: '12/01/1990',
      },
    ];

    const requestDescriptor = getUsersIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUsersIterator.next(users).value;
    expect(putDescriptor).toEqual(put(slice.actions.usersLoaded(users)));
  });

  it('should dispatch the user not found error', () => {
    username = 'test';

    const requestDescriptor = getUsersIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUsersIterator.throw({
      response: { status: 404 },
    }).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.repoError(RepoErrorType.USER_NOT_FOUND)),
    );
  });
  it('should dispatch the user has no repo error', () => {
    username = 'test';
    users = [];

    const requestDescriptor = getUsersIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUsersIterator.next(users).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.repoError(RepoErrorType.USER_HAS_NO_REPO)),
    );
  });
  it('should dispatch the github rate limit error', () => {
    username = 'test';

    const requestDescriptor = getUsersIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUsersIterator.throw(
      new Error('Failed to fetch'),
    ).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.repoError(RepoErrorType.GITHUB_RATE_LIMIT)),
    );
  });

  it('should dispatch the response error', () => {
    username = 'test';

    const requestDescriptor = getUsersIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUsersIterator.throw(new Error('some error')).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.repoError(RepoErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('githubRepoFormSaga Saga', () => {
  const githubRepoFormIterator = usersSaga();
  it('should start task to watch for loadRepos action', () => {
    const takeLatestDescriptor = githubRepoFormIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.actions.loadUsers.type, getUsers),
    );
  });
});
