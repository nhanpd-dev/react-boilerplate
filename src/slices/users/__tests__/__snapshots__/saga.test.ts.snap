// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUsers Saga should dispatch the github rate limit error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should dispatch the github rate limit error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getUsers Saga should dispatch the github rate limit error 3`] = `undefined`;

exports[`getUsers Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getUsers Saga should dispatch the reposLoaded action if it requests the data successfully 3`] = `undefined`;

exports[`getUsers Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getUsers Saga should dispatch the response error 3`] = `undefined`;

exports[`getUsers Saga should dispatch the user has no repo error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should dispatch the user has no repo error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getUsers Saga should dispatch the user has no repo error 3`] = `undefined`;

exports[`getUsers Saga should dispatch the user not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should dispatch the user not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getUsers Saga should dispatch the user not found error 3`] = `undefined`;

exports[`getUsers Saga should return error if username is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUsers Saga should return error if username is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": undefined,
      "type": "users/usersLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
